### DISCUSSION TITLE
host code撰寫時在vitis中無法抓取到kernel內的sub-function

### DISCUSSION THREAD

**Discussion by ZheChen-Bill** [bol-edu/HLS-SOC-Discussions#50] [General]
在撰寫final project時，我們的kexp2 kernel中會使用到sbox的kernel，因此我們將此二kernel import進入vitis的kernel
![圖片](https://github.com/bol-edu/HLS-Discussions/assets/88698677/ff6ba0e4-271b-4164-b5cf-f67df8916510)
我們的kernel -> hardware-link -> host -> system 4個步驟都能正確的build
![圖片](https://github.com/bol-edu/HLS-Discussions/assets/88698677/1a7b425d-ba0f-4aba-a2b0-036bb80c7fb5)

然而，在執行run configurations則出現以下錯誤:
![圖片](https://github.com/bol-edu/HLS-Discussions/assets/88698677/cc486671-590e-4ddb-8930-f875e9122a52)

似乎在執行cl::Program時無法讀取kernel裡使用的sub-function並寫入硬體中
我們已經上網找尋許多資料都無法解決這個問題，想請問有沒有方法可以解決這個問題

**Comment by boledulab** [comment]
In Lab3 - OpenCL/XRT lab, there are multiple kernel examples. Do you follow the same flow? I will suggest you use lab3 code as a template and stuff in your design.

**Comment by ZheChen-Bill** [comment]
我們有按照lab3的步驟執行，但是我們code的結構上與lab3的code不太相同，lab3的code是每個KA,KB等function都是top kernel，最後由host code去分別call這些kernel，而我們上述圖片的kexp2和sbox則是kexp2是top kernel，在host code執行時會只使用到kexp2，而kexp2會再去使用sbox這個sub-function

**Comment by boledulab** [comment]
I don't know if you wrap the function with an extern.  refer to  https://xilinx.github.io/Vitis-Tutorials/2022-1/build/html/docs/Getting_Started/Vitis/Part3.html 
Vitis requires C++ kernels to be declared as extern “C” to avoid name mangling issues

**Comment by ZheChen-Bill** [comment]
我們有使用extern "C" 的方式去標示主要的kernel 但是最後結果還是讀取不到 所以我們最後還是採用把所有使用到的sub-function kernel都放置在main function 的code檔案中才能執行run configuration
